capa/ida/plugin/proxy.py

```diff
@@ -110,3 +111,85 @@ def reset_address_range_filter(self):
         self.min_ea = None
         self.max_ea = None
         self.invalidateFilter()
+
+
+class CapaExplorerSearchProxyModel(QtCore.QSortFilterProxyModel):
+    """A SortFilterProxyModel that accepts rows with a substring match for a configurable query.
+
+    Looks for matches in the RULE_INFORMATION column (e.g. column 0).
+    Displays the entire tree row if any of the tree branches,
+     that is, you can filter by rule name, or also
+     filter by "characteristic(nzsor)" to filter matches with some feature.
+    """
+    def __init__(self, parent=None):
+        """ """
+        super(CapaExplorerSearchProxyModel, self).__init__(parent)
+        self.query = ""
+        self.setFilterKeyColumn(CapaExplorerDataModel.COLUMN_INDEX_RULE_INFORMATION)
+        self.setDynamicSortFilter(True)
+
+    def filterAcceptsRow(self, row, parent):
+        """true if the item in the row indicated by the given row and parent
+        should be included in the model; otherwise returns false
+
+        @param row: int
+        @param parent: QModelIndex*
+
+        @retval True/False
+        """
+        # this row matches, accept it
+        if self.filter_accepts_row_self(row, parent):
+            return True
+
+        # the parent of this row matches, accept it
+        alpha = parent
+        while alpha.isValid():
+            if self.filter_accepts_row_self(alpha.row(), alpha.parent()):
+                return True
+            alpha = alpha.parent()
+
+        # this row is a parent, and a child matches, accept it
+        if self.index_has_accepted_children(row, parent):
+            return True
+
+        return False
+
+    def index_has_accepted_children(self, row, parent):
+        """returns True if the given row or its children should be accepted"""
+        source_model = self.sourceModel()
+        model_index = source_model.index(row, 0, parent)
+
+        if model_index.isValid():
+            for idx in range(source_model.rowCount(model_index)):
+                if self.filter_accepts_row_self(idx, model_index):
+                    return True
+                if self.index_has_accepted_children(idx, model_index):
+                    return True
+
+        return False
+
+    def filter_accepts_row_self(self, row, parent):
+        """returns True if the given row should be accepted"""
+        if self.query == "":
+            return True
+
+        source_model = self.sourceModel()
+
+        index = source_model.index(row, 0, parent)
+        data = source_model.data(index, Qt.DisplayRole)
```

Comment:
@mike-hunhoff your existing code using `internalPointer()` didn't work for me here... not sure if your approach is needed for a reason i don't understand, or if we can do your filter in a more idiomatic way.

Reply:
ah, that's because the when using a `ProxyModel.index(...)`, this returns the item from the direct ancestor source model. if there are chained ProxyModels, it does not resolve the item from the CapaExplorerDataModel. this means that the internal pointer is not valid until the item has been resolved all the way up to the CapaExplorerDataModel.

Reply:
see changes here, which fix a similar problem in a different part of the code: https://github.com/fireeye/capa/pull/285/commits/e8cf19caf4043230acf8e450950ce5b172f7d6b8